test/robustness/overlay/areal_areal/random_integer_grids.cpp

```diff
@@ -0,0 +1,267 @@
+        }
+    }
+    return out;
+}
+
+std::ostream& operator<<(std::ostream& os, std::pair<bits, grid_settings> const& b_gs)
+{
+    os << '\n';
+    for (int y = b_gs.second.height - 1; y >= 0; --y)
+    {
+        for (int x = 0; x < b_gs.second.width; ++x)
+        {
+            int index = y * b_gs.second.width + x;
+            os << b_gs.first[index / chunk_size][index % chunk_size];
+        }
+        os << '\n';
+    }
+    return os;
+}
+
+bits geometry_to_bits(mp const& g, std::vector<point> const& tp)
+{
+    bits out((tp.size() + chunk_size - 1) / chunk_size);
+    for (size_t i = 0; i < tp.size(); ++i)
+    {
+        out[i / chunk_size][i % chunk_size] = bg::within(tp[i], g);
+    }
+    return out;
+}
+
+mp bits_to_geometry(bits const& b, std::vector<box> const& grid, std::vector<point> const& points,
+                    grid_settings const& settings, bool& all_success)
+{
+    mp out;
+    for (size_t i = 0; i < grid.size(); ++i)
+    {
+        if( b[i / chunk_size][i % chunk_size] )
+        {
+            mp temp;
+            bg::union_(out, grid[i], temp);
+            out = std::move(temp);
+        }
+    }
+    // Convenience lambda to pair bits with settings to use width/height in operator<<(os, ...)
+    const auto b_gs = [&settings](bits const& b) { return std::make_pair(b, settings); };
+    if ( ! bg::is_valid(out) )
+    {
+        std::cout << "Generating grid from pattern" << b_gs(b)
+                  << "results in invalid geometry: " << bg::wkt(out) << ".\n\n";
+        all_success = false;
+    }
+    if (geometry_to_bits(out, points) != b)
+    {
+        std::cout << "Generating grid from pattern" << b_gs(b)
+                  << "results in mismatching geometry: " << bg::wkt(out) << ".\n\n";
+        all_success = false;
+    }
+    return out;
+}
+
+bits gen_bits(prng& gen, int bits_size)
+{
+    bits b((bits_size + chunk_size - 1) / chunk_size);
+    std::generate(b.begin(), b.end(), std::ref(gen));
+    if (bits_size % chunk_size != 0)
+    {
+        std::bitset<chunk_size> bm;
+        bm.set();
+        bm >>= chunk_size - bits_size % chunk_size;
+        b.back() &= bm;
+    }
+    return b;
+}
+
+template <typename BitOp>
+bits apply_for_each(bits a, bits const& b, BitOp const& bit_op)
+{
+    for (size_t i = 0; i < a.size(); ++i) a[i] = bit_op(a[i], b[i]);
+    return a;
+}
+
+template<typename BitOp, typename GeoOp>
+void test_op(bits const& bits1, bits const& bits2, mp const& geo1, mp const& geo2,
+             std::string const& op_label, BitOp const& bit_op, GeoOp const& geo_op,
+             std::vector<point> const& points, std::vector<box> const& grid,
+             grid_settings const& settings, bool& success)
+{
+    auto test_geo = geo_op(geo1, geo2);
+    // Convenience lambda to pair bits with settings to use width/height in operator<<(os, ...)
+    const auto b_gs = [&settings](bits const& b) { return std::make_pair(b, settings); };
+    if ( ! bg::is_valid(test_geo) )
```

Comment:
You can get the reason (for invalidity) and output it, it is useful in these kind of programs